src/Interpreters/DatabaseCatalog.cpp

```diff
@@ -1046,9 +1046,13 @@ String DatabaseCatalog::getPathForDroppedMetadata(const StorageID & table_id) co
 
 String DatabaseCatalog::getPathForMetadata(const StorageID & table_id) const
 {
-    return getContext()->getPath() + "metadata/" +
-           escapeForFileName(table_id.getDatabaseName()) + "/" +
-           escapeForFileName(table_id.getTableName()) + ".sql";
+    auto database = getDatabase(table_id.getDatabaseName());
+    auto * database_ptr = dynamic_cast<DatabaseOnDisk *>(database.get());
+
+    if (!database_ptr)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Failed to get metadata path from database {}", table_id.getDatabaseName());
+
+    return database_ptr->getMetadataPath() + escapeForFileName(table_id.getTableName()) + ".sql";
```

Comment:
We should probably add another safeguard checking `database_ptr->getMetadataPath()` is not empty, in case somebody calls it for a database without metadata path (such as memory). Does it make sense?

Reply:
`getMetadataPath()` is only defined in DatabaseOnDisk class and DatabaseMemory is not inherited from it, but I added the check just in case.

Reply:
It is originally defined in `IDatabase` as returning an empty string, so all engines have the function (IDatabase -> DatabaseWithOwnTablesBase -> DatabaseMemory in the case of memory)